components/automate-ui/src/app/modules/infra-proxy/cookbook-details/cookbook-details.component.ts

```diff
@@ -0,0 +1,150 @@
+export type CookbookDetailsTab = 'details' | 'content';
+
+@Component({
+  selector: 'app-cookbook-details',
+  templateUrl: './cookbook-details.component.html',
+  styleUrls: ['./cookbook-details.component.scss']
+})
+export class CookbookDetailsComponent implements OnInit, OnDestroy {
+  private isDestroyed = new Subject<boolean>();
+  public cookbook: CookbookVersions;
+  public url: string;
+  public serverId: string;
+  public orgId: string;
+  public cookbookName: string;
+  public cookbookDetails: CookbookDetails;
+  public tabValue: CookbookDetailsTab = 'details';
+  public readFile: RootFiles;
+  public readFileUrl: string;
+  public readFileContent;
+  public cookbookDetailsLoading = false;
+  public cookbookVersionsLoading = true;
+  constructor(
+    private store: Store<NgrxStateAtom>,
+    private layoutFacade: LayoutFacadeService,
+    private http: HttpClient
+  ) { }
+
+  ngOnInit() {
+
+    this.layoutFacade.showSidebar(Sidebar.Infrastructure);
+    this.store.select(routeURL).pipe()
+      .subscribe((url: string) => {
+        this.url = url;
+        const [, fragment] = url.split('#');
+        this.tabValue = (fragment === 'content') ? 'content' : 'details';
+      });
+
+    combineLatest([
+      this.store.select(routeParams).pipe(pluck('id'), filter(identity)),
+      this.store.select(routeParams).pipe(pluck('orgid'), filter(identity)),
+      this.store.select(routeParams).pipe(pluck('cookbook_name'), filter(identity))
+    ]).pipe(
+      takeUntil(this.isDestroyed)
+    ).subscribe(([server_id, org_id, cookbook_name]) => {
+      this.serverId = server_id;
+      this.orgId = org_id;
+      this.cookbookName = cookbook_name;
+      this.store.dispatch(new GetCookbookVersions({
+        server_id: server_id,
+        org_id: org_id,
+        cookbook_name: cookbook_name
+      }));
+    });
+
+    combineLatest([
+      this.store.select(getStatus),
+      this.store.select(cookbookVersionsFromRoute)
+    ]).pipe(
+      filter(([getCookbookVersionSt, _cookbookVersionState]) =>
+        getCookbookVersionSt === EntityStatus.loadingSuccess),
+      filter(([_getCookbookVersionSt, cookbookVersionState]) =>
+        !isNil(cookbookVersionState)),
+      takeUntil(this.isDestroyed))
+      .subscribe(([_getCookbookVersionSt, cookbookVersionState]) => {
+        this.cookbook = cookbookVersionState;
+        this.cookbookVersionsLoading = false;
+        this.cookbookDetailsLoading = true;
+        this.store.dispatch(new GetCookbookDetails({
+          server_id: this.serverId,
+          org_id: this.orgId,
+          cookbook_name: this.cookbook.name,
+          cookbook_version: this.cookbook.versions[0]
+        }));
+        this.onCookbookVersionChange(
+          this.serverId,
+          this.orgId,
+          this.cookbook.name,
+          this.cookbook.versions[0]
+        );
+      });
+  }
+
+  public onCookbookVersionChange(
```

Comment:
We have a few issues to address in this method.

First, other than in exceptional circumstances, subscriptions must be in `ngOnInit` (or sometimes in the constructor). This is certainly not doing what you want: every time a cookbook version changes, you are creating a new subscription, accumulating potentially many of them, and all of them will run each time any version change occurs.

Second, such an `onXyz` method implies that on change, something is going to be done _now_. But this method does not actually do anything now; it only sets up a subscription that will do something some time in the future.

Third, this is nesting subscriptions three levels deep (one from the caller and two in this method)  which makes for very fragile code, and is ripe for hard-to-find bugs to occur.

Starting back on L88, you dispatch `GetCookbookDetails`. If I am interpreting the intent correctly, you should **not** be directly calling `onCookbookVersionChange` (L94). Rather, you should set up an independent (and unnested) subscription to monitor the ngrx store for `GetCookbookDetails` to complete and then do something which, I believe, should be similar to the body of this function.

Reply:
Thanks for pointing out the issues, I should have consider it before but some how I went off track :)
1. All the subscription are in `ngOnInit` now.
2. Removed extra method.
3. Added subscription to monitor the ngrx store.

Reply:
Thank you for those updates!
So now back to the question of tests. What I would suggest is just a few tests like these:
* Upon initializing, GetCookbookVersions is dispatched.
* Upon <something else>, GetCookbookDetails is dispatched.
You would do those by dispatching the `xyzSuccess` action or equivalent, as I think I mentioned previously.

For your last subscription block there is more complex logic there so perhaps a couple tests. You want to dispatch an action to trigger the subscription, as above, but then you need to manipulate the data so that the nested http.get gets called in one test and does not get called in another test. Hmm... not sure if we can mock the result of the http.get like we can mock the ngrx actions, though.

You can see some examples of "mocking" the ngrx actions in my recent PRs: #3562 and #3588.

Reply:
@msorens updated some tests, I think which will cover all above scenarios please have a look.
 :)